### PR TITLE
Make -dp handle folders as: include folder and all .?ar files inside.

### PR DESCRIPTION
**Pull request by npetzall**
* Refactored and split DbDriverLoader into 3 separate classes, since
  DbDriverLoaded resolved path, loaded driver and returned a connection.
* Created more specific configuration classes since one hugh has lots of
  concerns.
* Delegate classes for config since we are moving from one to another.
* Put in a new db package since we might want to split schemaspy later on.
* Updated documentation

### CODE REVIEW

**Comment by npetzall** [comment]
If this is totally wrong just decline it and I'll implement it in the DbDriverLoader.

Thing I'm lacking right now is the logging from DbDriverLoader

Line 47 if it's important I'll refactor Config class to actually log it, when it has determined dbProperties. I'll also switch to SLF4J in that case.

As for the logging in line 70 to 95 I could improve the documentation instead.

**Comment by rafalkasa** [comment]
@npetzall good job, you are the Hero 👍 I need to review your change, but after my first overview looks great. I try to find the time for detailed review and let you know my feedback.

Thank you for your time and contribution to the project. This is amazing how this project evolve in last month's.

**Comment by npetzall** [comment]
I want to continue with moving the determinedbproperties to org.schemaspy.db.config.PropertiesResolver

Since it's not pure config more a resolver or similar function. But that's mostly because I see config as a value object.

**Comment by npetzall** [comment]
Things that I believe should be done if this type of refactoring is desired.
* Construct Adapater, Loader and ConfigDelegate in SchemaSpyConfig and have them injected in constructor.

This will mean that it's dependency injected and easier to mock, replace etc.

File: src/main/java/org/schemaspy/db/driver/PathResolver.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                    return true;
+                })
+                .flatMap(addChildPaths)
+                .map(pathToUrl)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+        if (paths.isEmpty()) {
+            LOG.info("DriverPath resolved to: empty");
+        } else {
+            String strPaths = paths.stream().map(u -> u.getPath()).collect(Collectors.joining("\n\t", "\n\t", ""));
```
Comment (npetzall) [inline src/main/java/org/schemaspy/db/driver/PathResolver.java]: `\n` should be replaced with `System.lineSeparator()`

File: src/main/java/org/schemaspy/service/SqlService.java [hunk 2]
```diff
@@ -47,23 +49,7 @@ public DatabaseMetaData getMeta() {
...
-        if (driverPath == null)
-            driverPath = "";
-
-        if (config.getDriverPath() != null)
-            driverPath = config.getDriverPath();
-
-        DbDriverLoader driverLoader = new DbDriverLoader();
-        connection = driverLoader.getConnection(config, urlBuilder.build(), driverClass, driverPath);
-
+        connection = new Adapter(new AdapterConfigDelegate(commandLineArguments, config), new Loader(new LoaderConfigDelegate(commandLineArguments, config))).getConnection();
```
Comment (npetzall) [inline src/main/java/org/schemaspy/service/SqlService.java]: Adapter and config should either be injected thru Ctor or created in Ctor.

File: src/main/java/org/schemaspy/SchemaAnalyzer.java [hunk 3]
```diff
@@ -455,22 +459,7 @@ private void prepareLayoutFiles(File outputDir) throws IOException {
...
-        String driverClass = properties.getProperty("driver");
-        String driverPath = properties.getProperty("driverPath");
-        if (driverPath == null)
-            driverPath = "";
-        if (config.getDriverPath() != null)
-            driverPath = config.getDriverPath() + File.pathSeparator + driverPath;
-
-        DbDriverLoader driverLoader = new DbDriverLoader();
-        return driverLoader.getConnection(config, urlBuilder.build(), driverClass, driverPath);
+        return new Adapter(new AdapterConfigDelegate(commandLineArguments, config), new Loader(new LoaderConfigDelegate(commandLineArguments, config))).getConnection();
```
Comment (npetzall) [inline src/main/java/org/schemaspy/SchemaAnalyzer.java]: Injected in Ctor or created in Ctor atleast not like this.

File: src/main/java/org/schemaspy/db/driver/PathResolver.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        return paths;
+    }
+
+    private static Function<Path, Stream<Path>> addChildPaths = p -> {
+        if (Files.isDirectory(p)) {
+            try {
+                return Stream.concat(
+                        Stream.of(p),
+                        StreamSupport.stream(
+                                Files.newDirectoryStream(p, "*.?ar").spliterator(),
```
Comment (npetzall) [inline src/main/java/org/schemaspy/db/driver/PathResolver.java]: Possible resource leakage DirectoryStream should be closed.

**Comment by rafalkasa** [comment]
@npetzall please resolve the conflicts thx

**Comment by npetzall** [comment]
I started with the model refactoring and I found I have changed this too much. I think this should be closed and not merged.

**Comment by rafalkasa** [comment]
@npetzall based on your suggestion I will close this PR and not merge

**Pull request closed without merging**